File: app/src/main/java/comviewaquahp/google/sites/youbimiku/MainActivity.java
package comviewaquahp.google.sites.youbimiku;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Bundle;
import android.text.InputType;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.EditText;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.github.bassaer.chatmessageview.model.Message;
import com.github.bassaer.chatmessageview.view.ChatView;
import com.google.android.material.textfield.TextInputLayout;
import com.google.gson.Gson;
import com.google.gson.JsonElement;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import ai.api.AIServiceException;
import ai.api.RequestExtras;
import ai.api.android.AIConfiguration;
import ai.api.android.AIDataService;
import ai.api.android.GsonFactory;
import ai.api.model.AIContext;
import ai.api.model.AIError;
import ai.api.model.AIEvent;
import ai.api.model.AIRequest;
import ai.api.model.AIResponse;
import ai.api.model.Metadata;
import ai.api.model.Result;
import ai.api.model.Status;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    public static final String TAG = MainActivity.class.getName();
    private ChatView mChatView;
    private User masterAccount;
    private User mikuAccount;
    private AIDataService aiDataService;
    private Gson gson = GsonFactory.getGson();

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        SharedPreferences sharedPref = getPreferences(Context.MODE_PRIVATE);
        String userName = sharedPref.getString(getString(R.string.saved_user_name_key), "");

        initChatView();

        if (userName.isEmpty()) {
            showUserNameDialog(sharedPref);
        } else {
            masterAccount.setName(userName);
            showGreet(userName);
        }

        final LanguageConfig config =
                new LanguageConfig("ja", Constants.DIALOG_FLOW_ACCESS_TOKEN);
        initAIService(config);
    }

    @SuppressLint("ApplySharedPref")
    public void showUserNameDialog(SharedPreferences sharedPref) {
        final EditText editText = new EditText(this);
        editText.setInputType(InputType.TYPE_CLASS_TEXT);
        new AlertDialog.Builder(this)
                .setTitle(getString(R.string.tutorial_welcome))
                .setMessage(getString(R.string.tutorial_whats_your_name))
                .setView(editText)
                .setPositiveButton("OK", (dialog, which) -> {
                    SharedPreferences.Editor editor = sharedPref.edit();
                    String name = editText.getText().toString();
                    if (!name.isEmpty()) {
                        masterAccount.setName(name);
                    } else {
                        name = "";
                    }
                    showGreet(name);
                    editor.putString(getString(R.string.saved_user_name_key), name);
                    editor.commit();
                }).show();
    }

    @Override
    public void onClick(View v) {
        Message send = new Message.Builder()
                .setUser(masterAccount)
                .setRight(true)
                .setText(mChatView.getInputText())
                .hideIcon(true)
                .build();

        sendRequest(mChatView.getInputText());
        mChatView.send(send);
        mChatView.setInputText("");
    }

    private void onResult(final AIResponse response) {
        runOnUiThread(() -> {
            // Variables
            gson.toJson(response);
            final Status status = response.getStatus();
            final Result result = response.getResult();
            final String speech = result.getFulfillment().getSpeech();
            final Metadata metadata = result.getMetadata();
            final HashMap<String, JsonElement> params = result.getParameters();

            // Logging
            Log.d(TAG, "onResult");
            Log.i(TAG, "Received success response");
            Log.i(TAG, "Status code: " + status.getCode());
            Log.i(TAG, "Status type: " + status.getErrorType());
            Log.i(TAG, "Resolved query: " + result.getResolvedQuery());
            Log.i(TAG, "Action: " + result.getAction());
            Log.i(TAG, "Speech: " + speech);

            if (metadata != null) {
                Log.i(TAG, "Intent id: " + metadata.getIntentId());
                Log.i(TAG, "Intent name: " + metadata.getIntentName());
            }

            if (params != null && !params.isEmpty()) {
                Log.i(TAG, "Parameters: ");
                for (final Map.Entry<String, JsonElement> entry : params.entrySet()) {
                    Log.i(TAG, String.format("%s: %s",
                            entry.getKey(), entry.getValue().toString()));
                }
            }

            //Update view to bot says
            final Message receivedMessage = new Message.Builder()
                    .setUser(mikuAccount)
                    .setRight(false)
                    .setText(speech)
                    .build();
            mChatView.receive(receivedMessage);
        });
    }

    private void onError(final AIError error) {
        runOnUiThread(() -> Log.e(TAG, error.toString()));
    }

    /*
     * AIRequest should have query OR event
     */
    public void sendRequest(String text) {
        Log.d(TAG, text);
        if (TextUtils.isEmpty(text)) {
            onError(new AIError(getString(R.string.non_empty_query)));
            return;
        }

        new AiTask().execute(text, null, null);
    }

    private void initAIService(final LanguageConfig languageConfig) {
        final AIConfiguration.SupportedLanguages lang =
                AIConfiguration.SupportedLanguages.fromLanguageTag(languageConfig.getLanguageCode());
        final AIConfiguration config = new AIConfiguration(languageConfig.getAccessToken(),
                lang,
                AIConfiguration.RecognitionEngine.System);
        aiDataService = new AIDataService(this, config);
    }

    private void initChatView() {
        mChatView = findViewById(R.id.chat_view);
        mChatView.setMessageFontSize(Float.parseFloat(Constants.DEFAULT_MESSAGE_FONT_SIZE));
        mChatView.setUsernameFontSize(Float.parseFloat(Constants.DEFAULT_USERNAME_FONT_SIZE));
        mChatView.setTimeLabelFontSize(Float.parseFloat(Constants.DEFAULT_TIME_FONT_SIZE));

        Bitmap mikuFace = BitmapFactory.decodeResource(getResources(), R.drawable.normal);
        masterAccount = new User(0, null, null);
        mikuAccount = new User(1, getString(R.string.miku_name), mikuFace);

        mChatView.setOnClickSendButtonListener(this);
    }

    private void showGreet(String userName) {
        String greeting = getResources().getString(R.string.tutorial_nice_to_meet_you, userName);
        Message welcome = new Message.Builder()
                .setUser(mikuAccount)
                .setRight(false)
                .setText(greeting)
                .build();

        mChatView.receive(welcome);
    }

    public class AiTask extends AsyncTask<String, Void, AIResponse> {
        private AIError aiError;

        @Override
        protected AIResponse doInBackground(final String... params) {
            final AIRequest request = new AIRequest();
            String query = params[0];
            String event = params[1];
            String context = params[2];

            if (!TextUtils.isEmpty(query)) {
                request.setQuery(query);
            }

            if (!TextUtils.isEmpty(event)) {
                request.setEvent(new AIEvent(event));
            }

            RequestExtras requestExtras = null;
            if (!TextUtils.isEmpty(context)) {
                final List<AIContext> contexts = Collections.singletonList(new AIContext(context));
                requestExtras = new RequestExtras(contexts, null);
            }

            try {
                return aiDataService.request(request, requestExtras);
            } catch (final AIServiceException e) {
                aiError = new AIError(e);
                return null;
            }
        }

        @Override
        protected void onPostExecute(final AIResponse response) {
            if (response != null) {
                onResult(response);
            } else {
                onError(aiError);
            }
        }
    }
}

File: app/src/main/java/comviewaquahp/google/sites/youbimiku/MainActivity.kt
package comviewaquahp.google.sites.youbimiku

import android.content.Intent
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.util.DisplayMetrics
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.aallam.openai.api.completion.CompletionRequest
import com.aallam.openai.api.completion.TextCompletion
import com.aallam.openai.api.model.ModelId
import com.aallam.openai.client.OpenAI
import com.github.bassaer.chatmessageview.model.Message
import com.google.android.gms.ads.*
import com.google.android.gms.ads.admanager.AdManagerAdRequest
import com.google.android.gms.ads.admanager.AdManagerAdView
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.play.core.review.ReviewManagerFactory
import comviewaquahp.google.sites.youbimiku.config.*
import comviewaquahp.google.sites.youbimiku.databinding.ActivityMainBinding
import kotlinx.coroutines.*

class MainActivity : AppCompatActivity(), View.OnClickListener, DialogListener {
    private lateinit var userAccount: User
    private lateinit var mikuAccount: User

    private lateinit var binding: ActivityMainBinding
    private lateinit var detectIntent: DetectIntent
    private lateinit var adView: AdManagerAdView
    private lateinit var interstitialAd: InterstitialAd
    private lateinit var openAI: OpenAI

    private var openAIPreviousResponse = ""

    private var initialLayoutComplete = false
    private val job = SupervisorJob()
    private val exceptionHandler: CoroutineExceptionHandler =
        CoroutineExceptionHandler { value, throwable ->
            Log.e(TAG, throwable.message.toString())
        }
    private val scope =
        CoroutineScope(Dispatchers.Default + job + exceptionHandler) // exceptionHandlerを渡す

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)

        detectIntent = DetectIntent(this, getDialogFlowSession())

        initChatView()
        initBanner()

        showInAppReviewIfNeeded()

        openAI = OpenAI(BuildConfig.openAIKey)
        setup()

        val adRequest = AdRequest.Builder().build()
        InterstitialAd.load(this,BuildConfig.adInterstitialUnitId, adRequest, object : InterstitialAdLoadCallback() {
            override fun onAdFailedToLoad(adError: LoadAdError) {
                Log.d(TAG, adError.message)
            }
            override fun onAdLoaded(ad: InterstitialAd) {
                Log.d(TAG, "Ad was loaded.")
                interstitialAd = ad
            }
        })
    }

    private val adSize: AdSize
        get() {
            val display = windowManager.defaultDisplay
            val outMetrics = DisplayMetrics()
            display.getMetrics(outMetrics)

            val density = outMetrics.density

            var adWidthPixels = binding.adViewContainer.width.toFloat()
            if (adWidthPixels == 0f) {
                adWidthPixels = outMetrics.widthPixels.toFloat()
            }

            val adWidth = (adWidthPixels / density).toInt()
            return if (BuildConfig.FLAVOR == "ads") {
                AdSize.getCurrentOrientationAnchoredAdaptiveBannerAdSize(this, adWidth)
            } else {
                AdSize(0, 0)
            }
        }


    private fun initBanner() {
        MobileAds.initialize(this) {}
        MobileAds.setRequestConfiguration(
            RequestConfiguration.Builder().build()
        )

        binding.adPlaceholder.layoutParams.height = adSize.getHeightInPixels(this)
        adView = AdManagerAdView(this)
        binding.adViewContainer.addView(adView)
        binding.adViewContainer.viewTreeObserver.addOnGlobalLayoutListener {
            if (!initialLayoutComplete) {
                initialLayoutComplete = true
                loadBanner(adSize)
            }
        }
    }

    private fun loadBanner(adSize: AdSize) {
        Log.d(TAG, "loadBanner()")
        adView.adUnitId = BuildConfig.adBannerUnitId
        adView.setAdSizes(adSize)
        val adRequest = AdManagerAdRequest.Builder().build()
        adView.loadAd(adRequest)
    }

    private fun initChatView() {
        val size = FontSizeConfig.getSize(getFontSizeType(this))
        setFontSize(size, binding.chatView)

        userAccount = User(0, null, null)
        mikuAccount = getMikuAccount()
        val mlp = binding.chatView.layoutParams as ViewGroup.MarginLayoutParams
        mlp.topMargin = adSize.getHeightInPixels(this)
        binding.chatView.setDateSeparatorFontSize(0F)
        binding.chatView.setInputTextHint(getString(R.string.input_text_hint))
        binding.chatView.setOnClickSendButtonListener(this)
    }

    private fun getMikuAccount(): User {
        return if (getAIModel(this) == (AIModelConfig.OPEN_AI.name)) {
            val face = BitmapFactory.decodeResource(resources, R.drawable.glad)
            val name = "${getString(R.string.miku_name)}(GPT)"
            User(2, name, face)
        } else {
            val face = BitmapFactory.decodeResource(resources, R.drawable.normal)
            User(1, getString(R.string.miku_name), face)
        }
    }

    private fun showGreet(userName: String?) {
        val greeting = resources.getString(R.string.miku_nice_to_meet_you, userName)
        val welcome = Message.Builder()
            .setUser(mikuAccount)
            .setRight(false)
            .setText(greeting)
            .build()
        binding.chatView.receive(welcome)
    }

    private fun showOpenAIGreet(userName: String?) {
        val greeting = resources.getString(R.string.user_nice_to_meet_you, userName)
        scope.launch {
            openAITask(greeting)
        }
    }

    private fun showInAppReviewIfNeeded() {
        val pref = SharedPreferenceManager
        pref.get(this, Key.LAUNCH_COUNT.name, 0).let {
            val current = it + 1
            pref.put(this, Key.LAUNCH_COUNT.name, current)

            // 起動回数が5回のときにInAppReviewを表示しカウントをリセット
            if (current >= 5) {
                openInAppReview()
                pref.put(this, Key.LAUNCH_COUNT.name, 0)
            }
        }
    }

    private fun setup() {
        if (getUserName(this).equals("")) {
            showUserNameDialog(false)
        } else {
            userAccount.setName(getUserName(this).toString())
            when (getAIModel(this)) {
                AIModelConfig.OPEN_AI.name -> showOpenAIGreet(getUserName(this))
                else -> showGreet(getUserName(this))
            }
        }

        if (getAIModel(this).equals("")) {
            showAIModelDialog(false)
        }

    }

    private fun showUserNameDialog(cancelable: Boolean = true) {
        val dialog = UserNameDialogFragment()
        val args = Bundle()
        args.putBoolean(Constants.ARGUMENT_CANCELABLE, cancelable)
        dialog.arguments = args
        dialog.setDialogListener(this)
        dialog.show(supportFragmentManager, UserNameDialogFragment::class.java.name)
    }

    override fun doPositiveClick() {
        userAccount.setName(getUserName(this).toString())
        showGreet(getUserName(this))
    }

    private fun showAIModelDialog(cancelable: Boolean = true) {
        AlertDialog.Builder(this)
            .setTitle(getString(R.string.setting_ai_model))
            .setMessage(getString(R.string.setting_ai_model_message))
            .setPositiveButton(getString(R.string.setting_ai_model_openai)) { _, _ ->
                setAIModel(this, AIModelConfig.OPEN_AI)
                mikuAccount = getMikuAccount()

                interstitialAd.show(this)
            }
            .setNegativeButton(getString(R.string.setting_ai_model_dialogflow)) { _, _ ->
                setAIModel(this, AIModelConfig.DIALOG_FLOW)
                mikuAccount = getMikuAccount()
            }
            .setCancelable(cancelable)
            .show()
    }

    private fun showFontSizeDialog() {
        val index = FontSizeConfig.getType(getFontSizeType(this)).ordinal
        AlertDialog.Builder(this)
            .setTitle(getString(R.string.setting_font_size))
            .setSingleChoiceItems(R.array.font_size_config, index) { _, which ->
                setFontSize(FontSizeConfig.getSize(which), binding.chatView)
                SharedPreferenceManager.put(
                    this,
                    Key.FONT_SIZE.name,
                    FontSizeConfig.getType(which).name
                )
            }
            .setPositiveButton(getString(R.string.setting_dialog_accept), null)
            .show()
    }

    private fun showLanguageDialog() {
        val index = LanguageConfig.getType(getLanguage(this)).ordinal

        AlertDialog.Builder(this)
            .setTitle(getString(R.string.setting_language))
            .setSingleChoiceItems(R.array.language_config, index) { _, which ->
                SharedPreferenceManager.put(
                    this,
                    Key.LANGUAGE.name,
                    LanguageConfig.getType(which).name
                )
            }
            .setPositiveButton(getString(R.string.setting_dialog_accept), null)
            .show()
    }

    private fun openPlayStore() {
        try {
            val uri = Uri.parse("market://details?id=$packageName")
            val intent = Intent(Intent.ACTION_VIEW, uri)
            startActivity(intent)
        } catch (e: Exception) {
            Toast.makeText(
                this,
                getString(R.string.setting_submit_review_error),
                Toast.LENGTH_SHORT
            )
                .show()
        }
    }

    private fun openInAppReview() {
        try {
            val reviewManager = ReviewManagerFactory.create(this)
            reviewManager.requestReviewFlow().addOnSuccessListener { reviewInfo ->
                reviewManager.launchReviewFlow(this, reviewInfo)
                    .addOnSuccessListener {
                    }
            }
        } catch (e: Exception) {
        }
    }

    private fun openMailer() {
        try {
            val intent = Intent(Intent.ACTION_SENDTO).apply {
                data = Uri.parse("mailto:")
                putExtra(Intent.EXTRA_EMAIL, arrayOf(getString(R.string.feedback_email_address)))
                putExtra(Intent.EXTRA_SUBJECT, getString(R.string.feedback_email_subject))
                val text = buildString {
                    append("App Version: " + getVersionName())
                    append("\nModel Name: " + Build.MODEL)
                    append("\nOS Version: " + Build.VERSION.SDK_INT)
                    append("\n=================\n")
                    append(getString(R.string.feedback_email_inquiry))
                }
                putExtra(Intent.EXTRA_TEXT, text)
            }
            startActivity(intent)
        } catch (e: Exception) {
            Toast.makeText(
                this,
                getString(R.string.setting_send_feedback_error),
                Toast.LENGTH_SHORT
            )
                .show()
        }
    }

    private fun getVersionName(): String {
        return try {
            packageManager.getPackageInfo(packageName, 0).versionName
        } catch (e: Exception) {
            ""
        }
    }

    private fun openOfficialAccountIntent() {
        try {
            val uri = Uri.parse("https://twitter.com/youbimiku")
            val intent = Intent(Intent.ACTION_VIEW, uri)
            startActivity(intent)
        } catch (e: Exception) {
            Toast.makeText(
                this,
                getString(R.string.official_account_error),
                Toast.LENGTH_SHORT
            )
                .show()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.setting_user_name -> {
                showUserNameDialog()
                true
            }
            R.id.setting_ai_model -> {
                showAIModelDialog()
                true
            }
            R.id.setting_font_size -> {
                showFontSizeDialog()
                true
            }
            R.id.setting_language -> {
                showLanguageDialog()
                true
            }
            R.id.setting_submit_review -> {
                openPlayStore()
                true
            }
            R.id.setting_send_feedback -> {
                openMailer()
                true
            }
            R.id.setting_official_account -> {
                openOfficialAccountIntent()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onClick(v: View) {
        if (binding.chatView.inputText.isEmpty()) {
            return
        }
        val send = Message.Builder()
            .setUser(userAccount)
            .setRight(true)
            .setText(binding.chatView.inputText)
            .hideIcon(true)
            .build()
        sendRequest(binding.chatView.inputText)
        binding.chatView.send(send)
        binding.chatView.inputText = ""
    }

    private fun sendRequest(text: String) {
        Log.d(TAG, "request: $text")
        if (TextUtils.isEmpty(text)) {
            Log.e(TAG, "Text should not be empty.")
            return
        }

        var count = getOpenAIRequestCount(applicationContext)

        when (getAIModel(this)) {
            AIModelConfig.OPEN_AI.name ->
                scope.launch {
                    setOpenAIRequestCount(applicationContext, ++count)
                    openAITask(text)
                }
            else ->
                scope.launch {
                    dialogFlowTask(text)
                }
        }

        if (count >= OPENAI_REQUEST_COUNT_TO_SHOW_INTERSTITIAL) {
            interstitialAd.show(this)
            setOpenAIRequestCount(applicationContext, 0)
        }
    }

    private fun getDialogFlowSession(): String {
        val session = "youbimiku" + System.currentTimeMillis()
        Log.d(TAG, "getDialogFlowSession(): $session")
        return session
    }

    private suspend fun dialogFlowTask(text: String) {
        val response = detectIntent.send(text)
        Log.d(TAG, "response: $response")
        val receivedMessage = Message.Builder()
            .setUser(mikuAccount)
            .setRight(false)
            .setText(response)
            .build()
        withContext(Dispatchers.Main) {
            binding.chatView.receive(receivedMessage)
        }
    }

    private suspend fun openAITask(text: String) {
        val prompt = StringBuilder()
        if (openAIPreviousResponse.isNotEmpty()) {
            prompt.append("${getString(R.string.miku_name)}: ${openAIPreviousResponse}\n")
        }
        prompt.append("${userAccount.getName()}: ${text}\n${getString(R.string.miku_name)}:")
        Log.d(TAG, prompt.toString())
        val completionRequest = CompletionRequest(
            model = ModelId("text-davinci-003"),
            prompt = prompt.toString(),
            maxTokens = 256,
            echo = false,
            stop = listOf("\n")
        )
        val completion: TextCompletion = openAI.completion(completionRequest)
        val response = completion.choices.first().text.replace(" ", "")
        Log.d(TAG, "response: $response")
        openAIPreviousResponse = response
        val receivedMessage = Message.Builder()
            .setUser(mikuAccount)
            .setRight(false)
            .setText(response)
            .build()
        withContext(Dispatchers.Main) {
            binding.chatView.receive(receivedMessage)
        }
    }

    public override fun onPause() {
        adView.pause()
        super.onPause()
    }

    public override fun onResume() {
        super.onResume()
        adView.resume()
    }

    public override fun onDestroy() {
        adView.destroy()
        detectIntent.resetContexts()
        scope.coroutineContext.cancel()
        super.onDestroy()
    }

    companion object {
        val TAG = MainActivity::class.java.name.toString()
    }
}
